django/contrib/postgres/search.py

```diff
@@ -300,3 +303,90 @@ class TrigramSimilarity(TrigramBase):
 class TrigramDistance(TrigramBase):
     function = ''
     arg_joiner = ' <-> '
+
+
+class LexemeCombinable(Expression):
+    BITAND = '&'
+    BITOR = '|'
+
+    def _combine(self, other, connector, reversed, node=None):
+        if not isinstance(other, LexemeCombinable):
+            raise TypeError(
+                'Lexeme can only be combined with other Lexemes, '
+                'got {}.'.format(type(other))
+            )
+        if reversed:
+            return CombinedLexeme(other, connector, self)
+        return CombinedLexeme(self, connector, other)
+
+    # On Combinable, these are not implemented to reduce confusion with Q. In
+    # this case we are actually (ab)using them to do logical combination so
+    # it's consistent with other usage in Django.
+    def bitand(self, other):
+        return self._combine(other, self.BITAND, False)
+
+    def bitor(self, other):
+        return self._combine(other, self.BITOR, False)
+
+    def __or__(self, other):
+        return self._combine(other, self.BITOR, False)
+
+    def __and__(self, other):
+        return self._combine(other, self.BITAND, False)
+
+
+class Lexeme(LexemeCombinable, Value):
+    _output_field = SearchQueryField()
+
+    def __init__(self, value, output_field=None, *, invert=False, prefix=False, weight=None):
+        self.prefix = prefix
+        self.invert = invert
+        self.weight = weight
+        super().__init__(value, output_field=output_field)
+
+    def as_sql(self, compiler, connection):
+        param = "'%s'" % self.value.replace("'", "''").replace("\\", "\\\\")
```

Comment:
I am not 100% certain, but it may require to escape more than a single quote and backslash, specifically, all to_tsquery operators like & and |. The following should fail `Lexeme('red!')`. Here is the code I used to implement the escaping some time ago.
```

_SEARCH_SPEC_CHARS = r"['\0\[\]()|&:*!@<>\\]"
_spec_chars_re = re.compile(_SEARCH_SPEC_CHARS)
multiple_spaces_re = re.compile(r'\s{2,}')


def normalize_spaces(val):
    """Converts multiple spaces to single and strips from both sides"""
    if not val:
        return None
    return multiple_spaces_re.sub(' ', val.strip())

def psql_escape(query: str):
    # replace unsafe chars with space
    query = _spec_chars_re.sub(' ', query)
    query = normalize_spaces(query)  # convert multiple spaces to single 
    return query
```

Reply:
It looks like it should be handled by 'raw' search type, but it's not, and since quote and backslash is escaped, I assumed it should also handle other special characters.

Reply:
Thanks! I've updated it to use your code snippet for escaping

This did change to how some words are indexed though. For example: ``L'amour`` was previously indexed as ``L''amour``, and now it is ``L amour``. I'm not that familiar with French, so I'm not sure what impact that has.

Reply:
I've applied the suggestion from https://github.com/django/django/pull/12727#discussion_r528149294 which bought back the quotes, it also doesn't escape ``!``.

I think this is fine because the whole lexeme is surrounded by quotes, so your previous example would be escaped as ``'red!'`` and those quotes should prevent it from interpreting the ``!``.